File: problems/tower_of_hanoi.py

"""
solver for Tower of Hanoi problem
"""


class TowerOfHanoi(object):

    def __init__(self, size):

        self._size = size
        self._src = [i for i in range(size)]
        self._aux = list()
        self._dest = list()

    def solve(self):

        self._solve(self._size, self._src, self._aux, self._dest)

    def _solve(self, n, src, aux, dest):

        if n > 0:

            self._solve(n-1, src, aux, dest)

            self._dest.insert(0, src.pop())

            self._solve(n-1, aux, src, dest)

            print("-" * 5)
            print(src)
            print(aux)
            print(dest)
            print("-" * 5)


if __name__ == "__main__":

    tower = TowerOfHanoi(5)

    tower.solve()

File: problems/TowerOfHanoi.java
package program;

import java.util.Stack;

public class TheTowerOfHanoi {

	private static final String LINE = "------------------------";

	private Stack<Integer> src;
	private Stack<Integer> aux;
	private Stack<Integer> des;
	private int n;

	public static void main(String [] args){
		TheTowerOfHanoi tower = new TheTowerOfHanoi(5);
		tower.solve();
	}

	TheTowerOfHanoi(int n){

		this.n = n;
		src = new Stack<>();
		aux = new Stack<>();
		des = new Stack<>();

		for(int i=n;i>0;i--){

			src.add(i);

		}

	}


	void solve(){
		solve(n, src,aux,des);
	}

	void solve(int i,
	           Stack<Integer> src,
	           Stack<Integer> aux,
	           Stack<Integer> des){

		if (i > 0){

			solve(i-1, src, des, aux);

			des.push(src.pop());

			System.out.println(LINE);
			System.out.println(src);
			System.out.println(aux);
			System.out.println(des);
			System.out.println(LINE);

			solve(i-1, aux, src, des);

		}

	}

}